BJS2 - 15554 - Запрос на менторство

ravshanbekn opened the pull request:
Сделал часть 1-ю: Отправить запрос на менторство

ravshanbekn committed:
BJS2-15554 - Запрос на менторство

initial commit:
made classes, but did not implement logic yet

ravshanbekn committed:
BJS2-15554 - Запрос на менторство - Part-1

+ implemented required logic

pablo023 commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipRequestController.java:
```diff
@@ -0,0 +1,19 @@
+    private final MentorshipRequestValidator mentorshipRequestValidator;
+
+    public MentorshipRequestDto requestMentorship(MentorshipRequestDto mentorshipRequestDto) {
+        mentorshipRequestValidator.validateMentorshipRequestDescription(mentorshipRequestDto.getDescription());
```
Вот эта валидация должна осуществляться на уровне сервиса. На уровне контроллера ты валидируешь только корректность данных, которые идут в запросе - есть ли там нуллы, и прочее.

ravshanbekn replied:
Хорошо, можно еще валидировать минимальный id в контроллере? id >= MIN_ID?

pablo023 replied:
Нет, в контроллере не нужно валидаций каких-то, кроме как на поля, которые в классе dto аннотациями помечены. Есть момент с id интересный - там можно использовать класс-обертку, т.к. примитив если не указан, то он будет 0, а оберточный класс будет null, и это упадёт быстрее.

pablo023 commented on src/main/java/school/faang/user_service/dto/mentorship/MentorshipRequestDto.java:
Для лонгов можно использовать примитивы.

ravshanbekn replied:
Хорошо

shade2903 commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipRequestService.java:
```diff
@@ -0,0 +1,36 @@
+        return mentorshipRequestMapper.toDto(mentorshipRequest);
+    }
+
+    private void validateMentorshipRequest(long requesterId, long receiverId, LocalDateTime mentorshipCreationDate) {
```
Мне кажется в сервисе не должно быть реализаций отдельных методов валидаций, у тебя есть отдельный класс MentorshipRequestValidator и я бы этот метод туда перенес. А так в целом гуд

ravshanbekn replied:
Хорошо, спасибо

ravshanbekn committed:
BJS2-15554 - Запрос на менторство - Part-2

+ Сделал методы на получение и фильтрацию все запросов на менторство

ravshanbekn committed:
BJS2-15554 - Запрос на менторство - Part-3

+ Сделал методы на принятие запроса на менторство

pablo023 commented on src/test/java/school/faang/user_service/service/mentorship/MentorshipRequestServiceTest.java:
```diff
@@ -0,0 +1,250 @@
+class MentorshipRequestServiceTest {
+    @Captor
+    private ArgumentCaptor<MentorshipRequest> mentorshipRequestCaptor;
+    @Spy
```
Spy - плохая практика. Знаешь об этом?

ravshanbekn replied:
После дейлика знаю

ravshanbekn committed:
BJS2-15554 - Запрос на менторство - Part-4

+ Сделал методы для отклонения запроса на менторство

ravshanbekn committed:
BJS2-15554 - Запрос на менторство - Final Version

+ Отрефакторил, убрал @Spy, убрал не нужное

pablo023 commented on src/main/java/school/faang/user_service/validator/MentorshipRequestValidator.java:
```diff
@@ -0,0 +1,67 @@
+    public void validateMentorshipRequestDescription(String description) {
+        if (description == null || description.isBlank()) {
+            throw new IllegalArgumentException("There should be description");
+        } else if (description.length() > 4096) {
```
Чёт магией запахло)

ravshanbekn replied:
Исправлю

pablo023 commented on src/main/java/school/faang/user_service/validator/MentorshipRequestValidator.java:
```diff
@@ -0,0 +1,67 @@
+
+    public void validateMentorshipRequestReceiverAndRequesterExistence(long requesterId, long receiverId) {
+        if (!userRepository.existsById(requesterId)) {
+            throw new IllegalArgumentException("MentorshipRequest sender is not registered in Data Base");
```
Database*

ravshanbekn replied:
Хорошо

pablo023 commented on src/main/java/school/faang/user_service/validator/MentorshipRequestValidator.java:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+
+    public void validateReflection(long requesterId, long receiverId) {
```
Название такое себе. Reflection - это отдельная тема в Java,с которой тут создаётся ассоциация.

ravshanbekn replied:
Исправил

pablo023 commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipRequestService.java:
```diff
@@ -0,0 +1,82 @@
+        String description = mentorshipRequestDto.getDescription();
+        LocalDateTime mentorshipCreationDate = mentorshipRequestDto.getCreatedAt();
+
+        mentorshipRequestValidator.validateMentorshipRequestReceiverAndRequesterExistence(requesterId, receiverId);
+        mentorshipRequestValidator.validateReflection(requesterId, receiverId);
+        mentorshipRequestValidator.validateMentorshipRequestFrequency(requesterId, receiverId, mentorshipCreationDate);
```
Вот этот чейн валидаций как будто бы можно было бы в один метод запихать в валидаторе.

ravshanbekn replied:
Сделал

pablo023 commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipRequestService.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    public MentorshipRequestDto rejectRequest(long requestId, RejectionDto rejectionDto) {
+        MentorshipRequest mentorshipRequest = getMentorshipRequestByIdOrThrowException(requestId);
+        mentorshipRequestValidator.validateRequestStatusIsPending(mentorshipRequest.getStatus());
+        mentorshipRequest.setRejectionReason(rejectionDto.getRejectionReason());
+        setRequestStatusAndSaveToDataBase(mentorshipRequest, RequestStatus.REJECTED);
```
У тебя дублируется этот код в методах

ravshanbekn replied:
Вроде бы сделал, как то не уверен. Можешь еще раз посмотреть?

pablo023 commented on src/main/java/school/faang/user_service/validator/MentorshipRequestValidator.java:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+
+    public void validateMentorshipRequestReceiverAndRequesterExistence(long requesterId, long receiverId) {
```
validateMentorshipParticipants*

ravshanbekn replied:
Хорошо, исправил

pablo023 commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipRequestService.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    public MentorshipRequestDto acceptRequest(long requestId) {
+        MentorshipRequest mentorshipRequest = getMentorshipRequestByIdOrThrowException(requestId);
```
Не нужно создавать уровень абстракции, который просто одну строчку с эксепшеном заворачивает.

ravshanbekn replied:
Хорошо, исправил

pablo023 commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipRequestService.java:
```diff
@@ -0,0 +1,82 @@
+        MentorshipRequest mentorshipRequest = getMentorshipRequestByIdOrThrowException(requestId);
+        mentorshipRequestValidator.validateRequestStatusIsPending(mentorshipRequest.getStatus());
+        MentorshipRequest savedMentorshipRequest =
+                setRequestStatusAndSaveToDataBase(mentorshipRequest, RequestStatus.ACCEPTED);
```
Аналогично

ravshanbekn replied:
Исправил

pablo023 commented on src/main/java/school/faang/user_service/validator/MentorshipRequestValidator.java:
```diff
@@ -0,0 +1,67 @@
+    public void validateMentorshipRequestDescription(String description) {
+        if (description == null || description.isBlank()) {
+            throw new IllegalArgumentException("There should be description");
+        } else if (description.length() > 4096) {
```
Магия?)

ravshanbekn replied:
Исправил

pablo023 commented on src/main/java/school/faang/user_service/validator/MentorshipRequestValidator.java:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+
+    public void validateReflection(long requesterId, long receiverId) {
```
Reflection в Java имеет отдельный контекст. Название метода в заблуждение вводит

ravshanbekn replied:
Исправил

pablo023 reviewed (changes requested):
Поработай над комментами.

ravshanbekn committed:
BJS2-15554 - Запрос на менторство - Refactored Version

+ Исправил по комментариям

ravshanbekn committed:
BJS2-15554 - Запрос на менторство - Refactored Version

+ Отрефакторил фильтрацию и переписал тесты

pablo023 commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipRequestController.java:
```diff
@@ -0,0 +1,38 @@
+@RequiredArgsConstructor
+public class MentorshipRequestController {
+    private final MentorshipRequestService mentorshipRequestService;
+    private final MentorshipRequestValidator mentorshipRequestValidator;
```
Нам это здесь не нужно.

ravshanbekn replied:
Хорошо, исправил

pablo023 commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipRequestController.java:
```diff
@@ -0,0 +1,38 @@
+    private final MentorshipRequestValidator mentorshipRequestValidator;
+
+    public MentorshipRequestDto requestMentorship(MentorshipRequestDto mentorshipRequestDto) {
+        mentorshipRequestValidator.validateMentorshipRequestDto(mentorshipRequestDto);
```
Валидацию из контроллера убираем, это в сервис.

ravshanbekn replied:
Хорошо

pablo023 commented on src/main/java/school/faang/user_service/service/mentorship/MentorshipRequestService.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    public MentorshipRequestDto acceptRequest(long requestId) {
+        MentorshipRequest mentorshipRequest = getMentorshipRequestAndValidateStatusIsPending(requestId);
+        MentorshipRequest savedMentorshipRequest =
+                setRequestStatusAndSaveToDataBase(mentorshipRequest, RequestStatus.ACCEPTED);
```
Тут очень длинные названия методов, и вот когда у тебя метод делает два каких-то действия - это не гуд. Типа setStatus и saveToDataBase. Не надо никогда объединять, если ты используешь and в названии метода - что-то ты неправильно делаешь. И не DataBase, а repository. Можно в принципе на будущее - просто save. Итак понятно, что в репозиторий.

ravshanbekn replied:
В предыдущем ревью ты отметил этот же участок кода за повторение кода. Получается, один принцип надо все таки нарушить?

ravshanbekn replied:
А не, все, реализовал

pablo023 commented on src/test/java/school/faang/user_service/controller/mentorship/MentorshipRequestControllerTest.java:
```diff
@@ -0,0 +1,74 @@
+    private MentorshipRequestController mentorshipRequestController;
+
+    @Test
+    public void testRequestMentorshipValidationExecution() {
+        MentorshipRequestDto mentorshipRequestDto = MentorshipRequestDto.builder()
+                .description("description").build();
+
+        mentorshipRequestController.requestMentorship(mentorshipRequestDto);
+        verify(mentorshipRequestValidator, times(1))
+                .validateMentorshipRequestDescription(mentorshipRequestDto.getDescription());
+    }
```
Вот этот лишний, получается, т.к. валдиацию мы убираем из контроллера.

ravshanbekn replied:
Хорошо

pablo023 commented on src/test/java/school/faang/user_service/controller/mentorship/MentorshipRequestControllerTest.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @Test
+    public void testGetRequestsServiceExecution() {
```
Это не очень правильно. Ты тестируешь судя по названию метод сервиса, а в тесте дергаешь контроллер. надо сервис дергать.

ravshanbekn replied:
А мне разве не нужно проверять дергается ли service внутри controller?

pablo023 commented on src/test/java/school/faang/user_service/filter/mentorship/MentorshipRequestReceiverFilterTest.java:
```diff
@@ -0,0 +1,67 @@
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@ExtendWith(MockitoExtension.class)
```
А она тебе тут нужна? Ты не мокаешь ничего же.

ravshanbekn replied:
Исправил

pablo023 commented on src/test/java/school/faang/user_service/service/mentorship/MentorshipRequestServiceTest.java:
```diff
@@ -0,0 +1,279 @@
+    }
+
+    @Test
+    public void testRequestMentorshipMapperToDtoExecution() {
```
Маппер не нужно тестить, это сторонняя либа.

ravshanbekn replied:
Выполнение маппера тоже не нужно тестить?

pablo023 commented on src/test/java/school/faang/user_service/service/mentorship/MentorshipRequestServiceTest.java:
```diff
@@ -0,0 +1,279 @@
+        );
+    }
+
+    private List<MentorshipRequest> getMentorshipRequestList() {
+        return List.of(
+                getMentorshipRequest(1L, 2L, "description", RequestStatus.PENDING),
+                getMentorshipRequest(1L, 3L, "testing", RequestStatus.REJECTED),
+                getMentorshipRequest(2L, 3L, "some", RequestStatus.ACCEPTED),
+                getMentorshipRequest(3L, 4L, "some", RequestStatus.PENDING)
+        );
+    }
+
+    private MentorshipRequest getMentorshipRequest(long requesterId, long receiverId,
+                                                   String description, RequestStatus requestStatus) {
+        return MentorshipRequest.builder()
+                .requester(User.builder().id(requesterId).build())
+                .receiver(User.builder().id(receiverId).build())
+                .description(description)
+                .status(requestStatus)
+                .build();
+    }
```
Их куда-то в конец, либо в beforeEach / beforeAll

ravshanbekn replied:
Исправил

pablo023 reviewed (changes requested):
В целом окей. Убери валидатор из контроллера, исправь некоторые тесты, используй в тестах аннотацию DisplayName, поправь методы сервиса, где я указал, и окей. В целом всё хорошо, но по мелочи можно доработать)

ravshanbekn committed:
BJS2-15554 - Запрос на менторство

+ used @ DisplayName on tests
+ used @ RequestStatus
+ used @ Valid annotation
+ added ExceptionHandler
+ enhanced Service with @ Transactional annotation

ravshanbekn committed:
BJS2-15554 - Запрос на менторство

+ fixed inaccuracy

ravshanbekn committed:
BJS2-15554 - Запрос на менторство

+ controller request style

pablo023 commented on src/main/java/school/faang/user_service/dto/mentorship/MentorshipRequestDto.java:
```diff
@@ -0,0 +1,34 @@
+public class MentorshipRequestDto {
+    private long id;
+
+    @NotNull(message = "Description should not be null value")
```
NotBlank было бы еще неплохо добавить

ravshanbekn replied:
Хорошо

pablo023 commented on src/main/java/school/faang/user_service/dto/mentorship/RejectionDto.java:
```diff
@@ -0,0 +1,18 @@
+@AllArgsConstructor
+@Builder
+public class RejectionDto {
+    @NotNull
```
NotBlank

ravshanbekn replied:
Хорошо

pablo023 reviewed (approved):
Молодец) Мерджим!
